types/reach__menu-button/index.d.ts

```diff
@@ -45,14 +46,14 @@ export type ResolvedMenuLinkProps<T> = T extends keyof JSX.IntrinsicElements
   ? JSX.IntrinsicElements[T]
   : T;
 
-export type ResolvedMenuLinkComponent<T> = T extends keyof JSX.IntrinsicElements
-  ? T
-  : React.ComponentType<T>;
+export type ResolvedMenuLinkComponent<T> =
+  | keyof JSX.IntrinsicElements
+  | React.ComponentType<T>;
 
 export type MenuLinkProps<
   T extends SupportedMenuLinkComponent
 > = ResolvedMenuLinkProps<T> & {
-  as?: string;
+  as?: ResolvedMenuLinkComponent<T>;
```

Comment:
> Why isn't as typed as `T`?

Because `as` accepts to type of values:

* A string containing an HTML element name (eg: a, br, div)
* A React component

> Does the current implementation catch issues with the following?

I'm not sure to understand your question...

Reply:
I'm saying that `<MenuItem as="a" wat={true} />` should fail since `wat` isn't an expected property for `a` elements. Similarly, `<MenuItem as="a" title={42} />` should fail since `title` should be a `string`.

I am possibly misunderstanding how the library is meant to be used though.

Reply:
Oh, you're right I forgot that point...

So how can I fix the previous implementation then? Do you understand why it is broken?